src/openclsim/critical_path/dependencies_from_simpy_step.py

```diff
@@ -1,40 +1,245 @@
+                    simpy.events.Timeout,
+                )
+                and self.step_logging_dataframe.loc[elem, "event_object"]._delay > 0
+            )
+
+            if relevant_timeout:
+                # relevant to SAVE
+                if last_seen is not None:
+                    DEPENDENCIES_OPENCLSIM.append((last_seen, elem))
+                last_seen = elem
+
+            for effect_this_elem in effects_this_elem:
+                logging.debug(f"Effect {effect_this_elem} from {effects_this_elem}")
+                __extract_openclsim_dependencies(
+                    tree_input, elem=effect_this_elem, last_seen=last_seen
+                )
+
+            return None
+
+        # get all relevant dependencies from the simpy dependencies,
+        # that is find how the timeouts depend on one another.
+        tree = copy.deepcopy(self.cause_effect_list)
+        while len(tree) > 0:
+            __extract_openclsim_dependencies(tree)
+            tree = [tup for tup in tree if tup[0] not in SEEN]
+
+        # get recorded activities and convert times to floats (seconds since Jan 1970)
+        recorded_activities_df = self.recorded_activities_df.copy()
+        recorded_activities_df.start_time = round(
+            recorded_activities_df.start_time.astype("int64") / 10**9, 4
+        )
+        recorded_activities_df.end_time = round(
+            recorded_activities_df.end_time.astype("int64") / 10**9, 4
+        )
+
+        # rename the dependencies from dependencies with e_id to dependencies with cp_activity_id
+        self.dependency_list = [
+            (
+                self._find_cp_act(dependency[0], recorded_activities_df),
+                self._find_cp_act(dependency[1], recorded_activities_df),
+            )
+            for dependency in DEPENDENCIES_OPENCLSIM
+        ]
+
+    def _find_cp_act(self, e_id, recorded_activities_df):
+        """
+        Get cp activity ID given a time-window and an activity ID.
+
+        Parameters
+        ----------
+        e_id : int
+            Execution id from simpy.
+        recorded_activities_df : pd.DataFrame
+            From self.get_recorded_activity_df().
+        """
+        activity_id = self.step_logging_dataframe.loc[e_id, "event_object"].value
+        end_time = round(self.step_logging_dataframe.loc[e_id, "t1"], 4)
+        matching_ids = recorded_activities_df.loc[
+            (
+                (recorded_activities_df.ActivityID == activity_id)
+                & (recorded_activities_df.end_time == end_time)
+            ),
+            "cp_activity_id",
+        ]
+        if len(set(matching_ids)) == 1:
+            cp_activity_id = matching_ids.iloc[0]
+        else:
+            raise ValueError(
+                f"No recorded activity found for"
+                f" activityID {activity_id} at (end)time {end_time}"
+            )
+        return cp_activity_id
+
+
+class AlteredStepEnv(simpy.Environment):
```

Comment:
not sure about name, maybe ``CriticalPathEnvironment``?

Reply:
--> docstring will explain already that the step function os overwritten
--> change ``Env`` to ``Environment`` to comply with the simpy naming convention  ``simpy.Environment``

Reply:
CriticalPathEnvironment is better indeed